### PR TITLE
EVM/VM create() Static Constructor Reworking

### PR DESCRIPTION
**Pull request by holgerd77**

### CODE REVIEW

**Comment by holgerd77** [comment]
Please wait, I have forgotten the VM part.

File: packages/vm/src/vm.ts [hunk 1]
```diff
@@ -78,107 +77,57 @@ export class VM {
...
-    this.events = new AsyncEventEmitter<VMEvents>()
-
-    this._opts = opts
-
-    if (opts.common) {
-      this.common = opts.common
-    } else {
+    // Add common, SM, blockchain, EVM here
+    if (opts.common === undefined) {
       const DEFAULT_CHAIN = Chain.Mainnet
```
Comment (acolytec3) [inline packages/vm/src/vm.ts:82]: Do we really need this?  It feels unnecessary to declare this constant instead of just passing `Chain.Mainnet` in the next line to the `common` constructor.

File: packages/evm/src/evm.ts [hunk 2]
```diff
@@ -138,47 +139,54 @@ export class EVM implements EVMInterface {
...
-    this._optsCached = opts
+    const bn128 = await initRustBN()
 
-    this.transientStorage = new TransientStorage()
-
-    if (opts.common) {
-      this.common = opts.common
-    } else {
+    if (opts.common === undefined) {
       const DEFAULT_CHAIN = Chain.Mainnet
```
Comment (acolytec3) [inline packages/evm/src/evm.ts:156]: As in `vm`, this feels duplicative.  Can we just pass `Chain.Mainet` to the `common` constructor?

**Review by acolytec3 — DISMISSED** [review]
Looks good in general though left a couple of notes aboute code that feels unnecessarily wordy.

**Comment by jochem-brouwer** [comment]
Have pushed blockchain `_init` method removal in https://github.com/ethereumjs/ethereumjs-monorepo/pull/3315/commits/47f235db5e81daf72940e14e55a80325c2d382cc (remove from this PR if we want to merge it without the removal of blockchain `_init` and proceed in new PR)

Client tests fail, but only 2: to run: `npx vitest run ./test/net/protocol/ethprotocol.spec.ts` (in client)

**Comment by holgerd77** [comment]
> npx vitest run ./test/net/protocol/ethprotocol.spec.ts

Very cool! 🥳 

I would cautiously think we can keep this `_init()` removal in here? Would be nice if we have the "full thing" merged and that's it. Will have a look at the failing test(s), maybe I can spot something.

I would target to do the breaking releases on Thursday then, should be realistic. Last chance to last-minute get some manageably sized local refactoring of stuff (EVM/VM?) in, if you have spotted something along the work. 😋

**Comment by holgerd77** [comment]
Ok. I've now traced this down to happen during the following step respectively along this trace:

```shell
Trace:
    at Common.setHardfork (/ethereumjs-monorepo/packages/common/src/common.ts:316:13)
    at Common.setHardforkBy (/ethereumjs-monorepo/packages/common/src/common.ts:451:10)
    at Blockchain.checkAndTransitionHardForkByNumber (/ethereumjs-monorepo/packages/blockchain/src/blockchain.ts:1346:17)
    at Function.create (/ethereumjs-monorepo/packages/blockchain/src/blockchain.ts:194:24)
    at Function.create (/ethereumjs-monorepo/packages/client/src/blockchain/chain.ts:165:8)
    at /ethereumjs-monorepo/packages/client/test/net/protocol/ethprotocol.spec.ts:122:19
    at runTest (file:///ethereumjs-monorepo/node_modules/@vitest/runner/dist/index.js:486:11)
    at runSuite (file:///ethereumjs-monorepo/node_modules/@vitest/runner/dist/index.js:594:15)
    at runSuite (file:///ethereumjs-monorepo/node_modules/@vitest/runner/dist/index.js:594:15)
    at runFiles (file:///ethereumjs-monorepo/node_modules/@vitest/runner/dist/index.js:645:5)
```

So the outer `chainCommon` from client is changing from `london` to `chainstart` along this blockchain init code in the client:

```ts
console.log(`chain common before: ${options.config.chainCommon.hardfork()}`)
    options.blockchain =
      options.blockchain ??
      (await Blockchain.create({
        db: new LevelDB(options.chainDB),
        common: options.config.chainCommon,
        hardforkByHeadBlockNumber: true,
        validateBlocks: true,
        validateConsensus,
        genesisState: options.genesisState,
        genesisStateRoot: options.genesisStateRoot,
      }))
    console.log(`chain common after: ${options.config.chainCommon.hardfork()}`)
```

I am not fully sure why this didn't happen like this before. I nevertheless thought longer about this and would say that this is the behavior which we want since for blockchain - where the `common` is dynamic (in contrast to tx, block) the `common` (e.g. also the `chainCommon` from client) should move along with the inner Blockchain object.

I have cross-checked in the VM, if this e.g. has side-effects for the case that we instantiate a VM with a certain `common` and with a default blockchain created inside the VM (so, side effect I feared was that the VM common switches over to `chainstart` as well in this case).

This seems not to be the case, have found the following test in the VM in `test/api/index.ts`:

```ts
it('should succeed when common is set to a supported hardfork', async () => {
    const common = new Common({ chain: Chain.Mainnet, hardfork: Hardfork.Byzantium })
    const vm = await VM.create({ common })
    assert.equal(vm.common.hardfork(), Hardfork.Byzantium)
  })
```

So this is basically the described scenario.

Long story short 🙂: I think it is therefore justified to adjust the failing test by adding e.g. an additional `common` for the tx or explicitly reset the `chainCommon`. Will do so and push.

**Comment by jochem-brouwer** [comment]
@holgerd77 Thanks for the investigation and the fix 😃!

**Comment by holgerd77** [comment]
> @holgerd77 Thanks for the investigation and the fix 😃!

Ok, yeah, have pushed, also had another look at the whole code changes, so from my side - if tests pass - we can merge this (cannot approve, started the PR)!

**Review by jochem-brouwer — DISMISSED** [review]
LGTM

Regarding any further small refactoring: at this point I do not have anything to add

**Comment by jochem-brouwer** [comment]
Never mind, CI fails. Does it pass for you locally?

**Comment by holgerd77** [comment]
Ah no, it does not, just tested. This is another test, didn't realized that there were 2 affected files.

**Review by jochem-brouwer — APPROVED** [review]
LGTM!

**Review by jochem-brouwer — APPROVED** [review]
LGTM!
